benches/dapol.rs

```diff
@@ -1,175 +1,350 @@
-use core::fmt::Debug;
-use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, SamplingMode};
-use dapol::{
-    utils::get_secret, Dapol, DapolNode, RangeProofPadding, RangeProofSplitting, RangeProvable,
-    RangeVerifiable,
-};
-use digest::Digest;
-use rand::{distributions::Uniform, thread_rng, Rng};
-use smtree::{
-    index::TreeIndex,
-    traits::{ProofExtractable, Rand, Serializable, TypeName},
-};
-use std::time::Duration;
-
-// CONSTANTS
-// ================================================================================================
+mod setup;
+
+use criterion::{criterion_group, criterion_main};
+use criterion::{BenchmarkId, Criterion};
+use iai_callgrind::{black_box, library_benchmark, library_benchmark_group, main};
+
+// use std::path::PathBuf;
+use std::str::FromStr;
 
-const TREE_HEIGHTS: [usize; 3] = [16, 24, 32];
-const NUM_LEAVES: [usize; 3] = [1024, 2048, 4096];
+use dapol::{EntityId, Height, InclusionProof, MaxThreadCount};
 
-// BENCHMARKS
+use setup::{NUM_USERS, TREE_HEIGHTS};
+
+// BENCHMARKS: CRITERION
 // ================================================================================================
 
-fn build_dapol(c: &mut Criterion) {
+fn bench_build_tree(c: &mut Criterion) {
```

Comment:
The 3 loops need to be inside each other so that we fill the vector space `<num_entities, max_thread_count, height>`, i.e.
```rust
for h in heights {
  for m in max_thread_counts {
    for n in num_entities_list {
      let tree = build_tree(); // measure_mem, measure_time
      tree.serialize(); // measure_serialized_file_size
      let proof = tree.gen_proof(); // measure_mem, measure_time
      proof.serialize(); // measure_serialized_file_size
      proof.verify(); // measure_mem, measure_time
    }
  }
}
```
But I don't think you can do multiple measurements per benchmark so I'm not sure how easy it is to turn this pseudo-code into criterion/iai. Lmk what your thoughts are (also regarding maybe using a custom measurement).

Also would be ideal if we don't have to duplicate tree builds because when we get >10M entities the builds take really long.

Reply:
The whole min-10-runs + the 3 nested loops might end up being crazy long to execute for entities >10M so maybe it's worth spacing out the entity numbers more for >10M

Reply:
I used a tuple to allow for multiple param measurements, and it seems to work. I've just done it on the compute time though. I nested the for loops as suggested, and the metrics I am getting for a tree height of 16 hover around 1.5–2 seconds (for up to max 255 threads and 30k users). Does this sound about right? I'll run benches for the bigger trees later maybe as well. For now, I'll tackle serialization and then generating proofs

Reply:
I am also going to add a couple of issues later this evening, if that's okay

Reply:
Also just wanted to note that not all `(height, num_entities, max_thread_count)` tuples will be valid so the bench should only be done for valid tuples. Things that make a tuple invalid:
- height not big enough for num_entities
- max_thread_count greater than machine parallelism
- num_entities too big for machine memory (not sure how one would check for this)

Reply:
> Does this sound about right (on my 8-core 2019 MB Pro)?
Seems too fast but maybe it's right. Just push your code and I'll have a look.

> For now, I'll tackle serialization and then generating proofs
Cool, maybe split that into another PR, so this one can just focus on Criterion for tree building (makes for easier reviewing)

Reply:
I used a tuple to allow for multiple param measurements, and it seems to work. I've just done it on the compute time though.